kernel/mlx5/src/lib.rs

```diff
@@ -0,0 +1,219 @@
+//! A mlx5 driver for a ConnectX-5 100GbE Network Interface Card.
+//! 
+//! Currently we only support reading the device PCI space, mapping the initialization segment,
+//! and setting up a command queue to pass commands to the NIC.
+//! 
+//! All information is taken from the Mellanox Adapters Programmer’s Reference Manual (PRM) [Rev 0.54],
+//! unless otherwise specified. 
+
+#![no_std]
+#![allow(dead_code)] //  to suppress warnings for unused functions/methods
+
+#[macro_use] extern crate log;
+extern crate alloc;
+extern crate spin;
+extern crate irq_safety;
+extern crate memory;
+extern crate pci; 
+extern crate owning_ref;
+extern crate nic_initialization;
+extern crate mellanox_ethernet;
+extern crate kernel_config;
+
+use spin::Once; 
+use alloc::{
+    vec::Vec,
+    boxed::Box
+};
+use irq_safety::MutexIrqSafe;
+use memory::{PhysicalAddress, MappedPages, create_contiguous_mapping};
+use pci::PciDevice;
+use owning_ref::BoxRefMut;
+use nic_initialization::{NIC_MAPPING_FLAGS, allocate_memory};
+use mellanox_ethernet::{
+    InitializationSegment, 
+    command_queue::{CommandQueueEntry, CommandQueue, CommandOpcode, ManagePagesOpmod, QueryPagesOpmod}
+};
+use kernel_config::memory::PAGE_SIZE;
+
+/// Vendor ID for Mellanox
+pub const MLX_VEND:           u16 = 0x15B3;
+/// Device ID for the ConnectX-5 NIC
+pub const CONNECTX5_DEV:      u16 = 0x1019;
+
+/// The singleton connectx-5 NIC.
+/// TODO: Allow for multiple NICs
+static CONNECTX5_NIC: Once<MutexIrqSafe<ConnectX5Nic>> = Once::new();
+
+/// Returns a reference to the NIC wrapped in a MutexIrqSafe,
+/// if it exists and has been initialized.
+pub fn get_mlx5_nic() -> Option<&'static MutexIrqSafe<ConnectX5Nic>> {
+    CONNECTX5_NIC.get()
+}
+
+/// Struct representing a ConnectX-5 network interface card.
+pub struct ConnectX5Nic {
+    /// Initialization segment base address
+    mem_base: PhysicalAddress,
+    /// Initialization Segment
+    init_segment: BoxRefMut<MappedPages, InitializationSegment>,
+    /// Command Queue
+    command_queue: CommandQueue,
+    /// Boot pages passed to the NIC. Once transferred, they should not be accessed by the driver.
+    boot_pages: Vec<MappedPages>,
+    /// Init pages passed to the NIC. Once transferred, they should not be accessed by the driver.
+    init_pages: Vec<MappedPages>,
+}
+
+
+/// Functions that setup the NIC struct.
```

Comment:
i'm not sure if `///` triple slash doc comments atop an implementation block actually appear anywhere. Does this throw a warning when building the documentation?

Reply:
No, and the comment appears at the top of the implementation part of the docs

Reply:
oh very cool! I didn't know about that, perhaps it's a new addition to rustdoc.